bootstrap: add `spack --bootstrap` option for accessing bootstrap store

tgamblin opened the pull request:
We can see what is in the bootstrap store with `spack find -b`, and you can clean it with `spack clean -b`, but we can't do much else with it, and if there are bootstrap issues they can be hard to debug.

We already have `spack --mock`, which allows you to swap in the mock packages from the command line. This PR introduces `spack -b` / `spack --bootstrap`, which runs all of spack with `ensure_bootstrap_configuration()` set. This means that you can run `spack -b find`, `spack -b install`, `spack -b spec`, etc. to see what *would* happen with bootstrap configuration, to remove specific bootstrap packages, etc. This will hopefully make developers' lives easier as they deal with bootstrap packages.

This PR also uses a `nullcontext` context manager. `nullcontext` has been implemented in several other places in Spack, and this PR consolidates them to `llnl.util.lang`, with a note that we can delete the function if we ever reqyire a new enough Python.

- [x] introduce `spack --bootstrap` option
- [x] consolidated all `nullcontext` usages to `llnl.util.lang`

tgamblin committed:
bootstrap: add `spack --bootstrap` option for accessing bootstrap store

We can see what is in the bootstrap store with `spack find -b`, and you can clean it with `spack
clean -b`, but we can't do much else with it, and if there are bootstrap issues they can be hard to
debug.

We already have `spack --mock`, which allows you to swap in the mock packages from the command
line. This PR introduces `spack -b` / `spack --bootstrap`, which runs all of spack with
`ensure_bootstrap_configuration()` set. This means that you can run `spack -b find`, `spack -b
install`, `spack -b spec`, etc. to see what *would* happen with bootstrap configuration, to remove
specific bootstrap packages, etc. This will hopefully make developers' lives easier as they deal
with bootstrap packages.

This PR also uses a `nullcontext` context manager. `nullcontext` has been implemented in several
other places in Spack, and this PR consolidates them to `llnl.util.lang`, with a note that we can
delete the function if we ever reqyire a new enough Python.

- [x] introduce `spack --bootstrap` option
- [x] consolidated all `nullcontext` usages to `llnl.util.lang`

haampie commented:
Can't we just turn the whole bootstrap context into an ordinary environment? Instead of doing all this with imperative code:

```python
    bootstrap_store_path = store_path()
    with spack.architecture.use_platform(spack.architecture.real_platform()):
        with spack.repo.use_repositories(spack.paths.packages_path):
            with spack.store.use_store(bootstrap_store_path):
                # Default configuration scopes excluding command line
                # and builtin but accounting for platform specific scopes
                config_scopes = _bootstrap_config_scopes()
                with spack.config.use_configuration(*config_scopes):
                    with spack_python_interpreter():
                        yield
    spec_str = 'clingo-bootstrap@spack+python'
    
    python_prefix = sys.exec_prefix
    external_python = spec_for_current_python()

    entry = {
        'buildable': False,
        'externals': [
            {'prefix': python_prefix, 'spec': str(external_python)}
        ]
    }

    with spack.config.override('packages:python::', entry):
        yield
```

can't we ultimately put all these things into a declarative config/packages/specs sections of a `spack.yaml` file? Then `spack --bootstrap` could work like `spack -e ...` and you could also do stuff like `spack --bootstrap config blame config` etc to figure out what spack actually does under the hood.

tgamblin commented:
you can do config blame already:

```console
(spackle):cmd> spack --bootstrap config blame packages
---                                                                         packages:
overrides-0                                                                   python:
overrides-0                                                                     buildable: False
overrides-0                                                                     externals:
overrides-0                                                                     - prefix: /Users/gamblin2/Workspace/spack/var/spack/environments/default/.spack-env/._view/rkmuajirxlix2ypkp5sz3ygd3r3wskip
overrides-0                                                                       spec: 'python@3.8 '
/Users/gamblin2/Workspace/spack/etc/spack/defaults/darwin/packages.yaml:17    all:
/Users/gamblin2/Workspace/spack/etc/spack/defaults/packages.yaml:18             compiler: [apple-clang, clang, gcc, intel, pgi, xl, nag, fj, aocc]
/Users/gamblin2/Workspace/spack/etc/spack/defaults/darwin/packages.yaml:23      providers:
/Users/gamblin2/Workspace/spack/etc/spack/defaults/packages.yaml:24               elf: [libelf, elfutils]
```

But yes, it's not as ideal.  I think we should consider moving this stuff to a special environment eventually but at the moment it's in a bootstrap store so we can customize the specific install location.  We could think about changing that later -- I actually had the same discussion with @alalazo today.

haampie commented:
Ah right, of course that works ^^. I meant to say: it would be nice to be able to look at the spack.yaml file to demystify what spack bootstrap actually configures.

In any case, the `spack --bootstrap` api could work whether we model things with a bootstrap context, or an actual environment with spack.yaml file, so this PR looks good to me.

alalazo commented:
> Can't we just turn the whole bootstrap context into an ordinary environment?

If we move `bootstrapping` into an environment and treat it like an ordinary environment it means people using bootstrapped software have implicitly an environment already active when using Spack. So we need to define what would mean having two or more stacked environments for this to be implemented.

The other thing is that we may need an environment for each spec needed by Spack. Right now there is just `clingo` but when we'll be bootstrapping multiple specs it may happen that some of them are already present on the system where Spack is running. I think we may want to use those instead of pulling in an entire environment.

alalazo commented:
> it would be nice to be able to look at the spack.yaml file to demystify what spack bootstrap actually configures.

I am currently trying hard to keep all the configuration information for bootstrapping into `bootstrap.py`. 

In any case I think it would be difficult to have a single `spack.yaml` - you need to cover for different architectures and different external python interpreters which may be located anywhere, not to mention that you may need to concretize for `rhel5` with specific compilers to get binaries. 

So if we move to an environment the most we can do I think is have a template and render it somewhere based on the configuration of the Spack instance that is currently running. In that case I think we can also render the same information right now outside of an environment.

haampie commented:
> In any case I think it would be difficult to have a single spack.yaml

I mean, we can run all the stuff like adding specs, discovering compilers, changing config etc from imperative code, but in the end persist it as a `spack.yaml`

tgamblin commented:
@haampie @alalazo @becker33 any objections to this feature?  I just now wanted to do:

```console
$ spack -b location -i clingo-bootstrap
```

Currently there is no great programmatic analog for that -- I guess there is `spack find  -bp clingo-bootstrap`, but you have to do some grepping there.

I think this option would be useful for manipulating the bootstrap store without a lot of extra special cases.

alalazo reviewed (changes requested):
Command update is missing:
```
spack commands --update-completion
```

alalazo reviewed (dismissed):
Wondering if we should remove `spack find -b` but that can be another PR

tgamblin commented:
I think there is an issue with using this with uninstall and I need to add tests so I'll get to that in a bit

alalazo commented on lib/spack/spack/test/llnl/util/tty/log.py:
```diff
@@ -19,7 +19,7 @@
 
 import llnl.util.tty.log
-from llnl.util.lang import uniq
+from llnl.util.lang import nullcontext, uniq
```
Nitpick comment, but in general I would be for fully qualified namespaces instead of importing names from sub-sub-modules. The reason is that fully qualified namespaces are less subject to "circular-dependency" kind of issues + they are a good indicator of where some refactor is needed (if we constantly have to fully qualify a long name everywhere, it likely means the function we use belongs to a "higher" level in the chain of modules).

tgamblin replied:
Ok, this is done.

alalazo reviewed (dismissed):
Not merging since it was mentioned https://github.com/spack/spack/pull/25601#issuecomment-1040998904 that tests are missing, but LGTM.

tgamblin commented:
@alalazo I addressed the import style -- can you take another look?

I do not think I can create more tests for the part of `main()` this deals with, but I did test it by hand, and I adjusted the way it was done to make the changes very straightforward.  Can you merge if you think it is ok?

scottwittenburg commented on lib/spack/spack/main.py:
```diff
@@ -856,9 +859,22 @@ def _main(argv=None):
+        bootstrap_context = bootstrap.ensure_bootstrap_configuration()
+
+    with bootstrap_context:
+        finish_parse_and_run(parser, cmd_name, env_format_error)
```
@tgamblin @alalazo Any chance this should be a `return` statement?  I'm suspicious because suddenly returning a value from `ci_rebuild` isn't resulting in that value turning up as the exit code of the spack command.

@zackgalbreath fyi

alalazo replied:
@scottwittenburg I think you're right, looking at the previous code it seems we should return

scottwittenburg replied:
Ok, I can make a PR (with a test to make sure we notice in the future).

tgamblin replied:
It most definitely should be a return -- nice catch.
